base/src/ast.rs

```diff
@@ -1,4 +1,5 @@
 //! Module containing the types which make up `gluon`'s AST (Abstract Syntax Tree)
+use std::fmt;
 use std::marker::PhantomData;
 use std::ops::Deref;
```

Comment:
Wondering why we don't gain `Expr::TAbs` and `Expr::TApp` variants? How does this work for example:

```
id : forall a. a -> a = \x : a -> x
test = id 1
``` 
work?

Would have assumed we needed to expand to something like:

```
id : forall a. a -> a = \x : Type -> \x : a -> x
test = id Int 1
```

Wondering how Purescript represents this 🤔

Reply:
So the type of `id` is with some pseudo rust `Type::Forall(["a"], "a -> a", None)`. When we refer to `id` later in the code we add a new "skolem scope", this fills in the `None` part with fresh variables `Type::Forall(["a"], "a -> a", Some([$1]))`, one for each parameter to the `forall`. This basically amounts saying from which scope a variable originated so that we can check for escaping variables later.

This is purescript gets away with a slightly simpler implementation as it only marks this scope with an integer which bears no relation to the type variables. And that works fine for checking the types but falls a bit short when it comes to inferring the specialized type for each expression (ie `id : Int -> Int` rather than `id : forall a , a -> a` in the example above). This is where `TAbs` etc could have fixed the problem but I found it complicated to represent :/ . Instead, I inspect the variables bound in the `Some` part of the forall after type inference has finished and if a variable there has been unified, it gets removed from the `Type::Forall`. So since `$1` would be unified with `Int` we end up removing the forall completely and we get the specialized type.

It seems to work fine at least :P

Reply:
Maybe I am over exaggerating the complexity of `TAbs` but I didn't want to get into how to compute the types of `TAbs` and as I figured the solution above should work I thought that would be simpler.